### PR TITLE
Restructure SharedPreferences usage

### PR DESCRIPTION
**Pull request by johnjohndoe**
# Description
- In this branch `SharedPreferences` usage and "Settings" are cleaned up.
  - All preferences are **encapsulated** in `SharedPreferencesRepository.kt` resp. `preferences.xml`.
  - The `AppRepository` acts as the **central** access location to read or write preferences from Activities, Fragments, BroadcastReceivers.
  - New: All preference **default values** are set at initial app launch.
  - Preference values in the "Settings" screen are **automatically** stored by the Android framework. Redundant storing method calls are removed.
  - Names of preference resource follow **common conventions** (starting with: `preference_key_`, `preference_default_value_`, `preference_title_`, ..).
  - Formerly mixed naming and usage of `BUNDLE_` and `PREFS_` constants is **cleaned up**.
  - In a few cases preference keys are changed **without migrating** their values.
  - `SettingsActivity` is converted to **Kotlin**.
- Please refer to the individual commit messages for further details.

# Test scenarios
1. Enter alternative schedule URL. Validate that URL is loaded automatically.
2. Remove alternative schedule URL. Validate that default URL is loaded automatically.
3. Toggle the alternative highlighting setting. Validate that schedule is redrawn.
4. Select default alarm time. Validate that selected time is pre-selected when setting up an alarm for a session.
5. Enter Engelsystem URL. Validate that URL is loaded automatically and shifts show up.
6. Remove Engelsystem URL. Validate that shifts still show up.
7. Install app before this branch. Choose custom settings. Update app with this branch. Validate that custom settings are present.

# Successfully validated with temporary debugging output
- Automatic updates
- Alarm tone
- Insistent alarms 

# Successfully tested on
with `ccc36c3` flavor, `debug` build
- :heavy_check_mark: Nexus 5 emulator, Android 4.3.1 (API 18)
- :heavy_check_mark: Pixel C x86 emulator, Android 4.4.2 (API 19)

### CODE REVIEW

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.java [hunk 1]
```diff
@@ -47,22 +44,20 @@ public void onReceive(Context context, Intent intent) {
...
 
             Intent launchIntent = MainActivity.createLaunchIntent(context, sessionId, day);
             PendingIntent contentIntent = PendingIntent
                     .getActivity(context, lid, launchIntent, PendingIntent.FLAG_ONE_SHOT);
 
             NotificationHelper notificationHelper = new NotificationHelper(context);
-            String defaultReminderTone = context.getString(R.string.preferences_reminder_tone_default_value);
-            Uri soundUri = Uri.parse(prefs.getString("reminder_tone", defaultReminderTone));
+            String alarmToneUriString = appRepository.readAlarmTone();
+            Uri soundUri = Uri.parse(alarmToneUriString);
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.java:55]: It looks like you do this at all call sites of `AppRepository.readAlarmTone()`. Why not return a `Uri` from the repository?

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/contract/BundleKeys.java [hunk 2]
```diff
@@ -39,24 +39,10 @@
             "nerd.tuxmobil.fahrplan.congress.ChangesDialog.VERSION";
 
     // Shared Preferences
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/contract/BundleKeys.java:41]: Looks like this comment should be removed now.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/preferences/SharedPreferencesRepository.kt [hunk 3]
```diff
@@ -5,55 +5,95 @@ import android.content.SharedPreferences
 import android.preference.PreferenceManager
 import androidx.core.content.edit
 import nerd.tuxmobil.fahrplan.congress.R
-import nerd.tuxmobil.fahrplan.congress.contract.BundleKeys
 
 class SharedPreferencesRepository(val context: Context) {
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/preferences/SharedPreferencesRepository.kt:9]: I'd name this something like `SettingsRepository`. The fact that `SharedPreferences` are used is an implementation detail that's not important when using the repository.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepository.kt [hunk 4]
```diff
@@ -390,30 +390,38 @@ object AppRepository {
         metaDatabaseRepository.insert(values)
     }
 
+    fun readAlarmTimeIndex() =
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepository.kt:393]: Why route everything through `AppRepository`? That'll make it even more of a god object. I'd make callers use `SettingsRepository` directly.

Why the `read` prefix? Why not `get`?

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/settings/SettingsActivity.kt [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+        if (savedInstanceState == null) {
+            fragmentManager.beginTransaction()
+                    .add(R.id.container, SettingsFragment())
+                    .commit()
+            MyApp.LogDebug(LOG_TAG, "onCreate fragment created")
+        }
+    }
+
+    class SettingsFragment : PreferenceFragment() {
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/settings/SettingsActivity.kt:52]: I'd move the fragment to a separate file.

File: app/src/main/res/values/preferences.xml [hunk 6]
```diff
@@ -7,42 +7,101 @@
     <!-- Category general -->
     <string name="preference_key_category_general" translatable="false">preference_key_category_general</string>
 
+    <!-- Automatic schedule updates -->
```
Comment (cketti) [inline app/src/main/res/values/preferences.xml:10]: Why mix constants and translatable UI strings in one file?
Why do UI strings for settings live in a separate file? Wouldn't a single file be much easier for translators?

File: app/src/main/res/xml/prefs.xml [hunk 7]
```diff
@@ -9,50 +9,50 @@
         android:title="@string/general_settings">
 
         <CheckBoxPreference
-            android:defaultValue="@bool/preferences_auto_update_enabled_default_value"
-            android:key="auto_update"
-            android:title="@string/pref_auto_download" />
+            android:defaultValue="@bool/preference_default_value_auto_update_enabled"
```
Comment (cketti) [inline app/src/main/res/xml/prefs.xml:12]: Personal opinion: Using string resources for key names and values makes looking up things very annoying. Not all duplication is bad.

**Review by cketti — APPROVED** [review]

File: app/src/main/res/values/preferences.xml (hunk 6 above)
Comment (johnjohndoe) [reply app/src/main/res/values/preferences.xml:10]: Thanks. I guess you referred to `preference_title_app_notification_settings`. I moved the strings in bc6bbd636d1a609e46d9cd17b6b8a9f8f9528d1c.

Comment (johnjohndoe) [reply app/src/main/res/values/preferences.xml:10]: I want to have all preference resources in a single place. The mixed location in the past turned out to introduce a lot of errors.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepository.kt (hunk 4 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepository.kt:394]: The `AppRepository` hides whether data is persisted in shared preferences, a database or a file from the app layer above. The methods exposed by the `SharedPreferencesRepository` should only be accessed from the `AppRepository` layer.

I have chosen `read` following the CRUD schema.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/preferences/SharedPreferencesRepository.kt (hunk 3 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/preferences/SharedPreferencesRepository.kt:9]: In the `AppRepository` layer I prefer to have this level of implementation detail.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/contract/BundleKeys.java (hunk 2 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/contract/BundleKeys.java]: Thanks. Renamed in 19389b35329244512c2d79c362a6d1eb9dea48c5.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.java (hunk 1 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.java]: I agree. Changed in 0ea7501fe86c818c94ee3f33b47e3d86175865da.

Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/alarms/AlarmReceiver.java]: I agree. Changed in 0ea7501fe86c818c94ee3f33b47e3d86175865da.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/settings/SettingsActivity.kt (hunk 5 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/settings/SettingsActivity.kt]: Done in 0035e5eac9984ae1eb073478ad87593f61e65dc3.

Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/settings/SettingsActivity.kt]: Done in 0035e5eac9984ae1eb073478ad87593f61e65dc3.

File: app/src/main/res/xml/prefs.xml (hunk 7 above)
Comment (johnjohndoe) [reply app/src/main/res/xml/prefs.xml:12]: This was actually one of the top reasons to refactor this. The former distributed keys caused problems over the years.

**Comment by johnjohndoe** [comment]
@cketti Done. Back to you.

**Comment by johnjohndoe** [comment]
@cketti Can you test on a tablet please? I currently do not have access to one.

**Comment by cketti** [comment]
> @cketti I am done with applying my changes.  Back to you.

Still LGTM.

> @cketti Can you test on a tablet please? I currently do not have access to one.

Sorry, I don't have time to test this in the foreseeable future.

**Comment by johnjohndoe** [comment]
Successfully tested on a tablet :heavy_check_mark: Pixel C x86 emulator, Android 4.4.2 (API 19)

**Pull request merged**